Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp

```diff
@@ -224,6 +225,8 @@ void DownloadProxy::didFinish()
     updateQuarantinePropertiesIfPossible();
 #endif
     m_client->didFinish(*this);
+    if (m_downloadIsCancelled)
+        return;
```

Comment:
It looks like the intended behavior here is that either `DownloadClient::didFinish` or `DownloadClient::didFail` gets called exactly once. But I'm not sure whether canceling triggers didFail to be called or not? Do you know if it does?

 * If it doesn't, then your change here is good.
 * But if it does, then we'll wind up with didFail being called and then didFinish also being called next, which is not desired. In this case, we should check `if (m_d ownloadIsCancelled)` immediately _before_ `m_client->didFinish` rather than after. And same change would be needed in `DownloadProxy::didFail` as well.
 
 Please check!

Reply:
It doesn't seem triggering didFail.